src/internal/consistenthashing/etcd.go

```diff
@@ -0,0 +1,370 @@
+}
+
+// New is the default constructor for Ring. Creating a Ring starts a goroutine to watch prefix in etcd and creates a
+// node that continuously refreshes a lease in etcd.
+func New(ctx context.Context, prefix string, client *etcd.Client, logger *logrus.Logger, id ...string) (*Ring, error) {
+	if id == nil {
+		id = []string{uuid.New()}
+	}
+
+	ring := defaultRing(ctx, prefix, client, logger)
+	if err := ring.start(id[0]); err != nil {
+		return nil, errors.EnsureStack(err)
+	}
+
+	return ring, nil
+}
+
+// GetIndex returns the index of the Member associated with key. A key belongs to the first Member whose hash is
+// greater than the key. If no Members are greater, then the Ring wraps around to the first Member.
+func (ring *Ring) GetIndex(key string) int {
+	keyHash := ring.hashFn([]byte(key))
+	index := sort.Search(len(ring.members), func(i int) bool {
+		return ring.members[i].Hash >= keyHash
+	})
+
+	// wrap around to the first node in the Ring.
+	if index >= len(ring.members) {
+		index = 0
+	}
+
+	return index
+}
+
+// Get is a convenience wrapper around GetIndex that returns the Member that associates to a key.
+func (ring *Ring) Get(key string) Member {
+	return ring.members[ring.GetIndex(key)]
+}
+
+// Lock attempts to lock a key in etcd using consistent hashing. If key does not belong to the ring's node,
+// the call blocks until context is cancelled or expires.
+func (ring *Ring) Lock(ctx context.Context, key string) (context.Context, error) {
+	ticker := time.NewTicker(time.Second * 5)
+	defer ticker.Stop()
+	for {
+		index := ring.GetIndex(key)
+		if ring.node.Id == ring.members[index].Id {
+			break
+		}
+
+		select {
+		case <-ctx.Done():
+			return nil, errors.EnsureStack(ctx.Err())
+		case <-ticker.C:
+		}
+	}
+
+	if info, alreadyExists := ring.node.Locks.Load(key); alreadyExists {
+		lockCtx := info.(lockInfo).Ctx
+		return lockCtx, nil
+	}
+
+	nodeLock := dlock.NewDLock(ring.Client, key)
+	lockCtx, err := nodeLock.Lock(ctx)
+	if err != nil {
+		return nil, errors.EnsureStack(err)
+	}
+
+	ring.node.Locks.Store(key, lockInfo{
+		Lock: nodeLock,
+		Ctx:  lockCtx,
+	})
+
+	ring.Logger.WithFields(ring.lockLogFields(key)).Info("claimed lock")
+
+	return lockCtx, nil
+}
+
+// Unlock attempts to unlock a key in etcd if and only if a ring's node owns a lock reference to it.
+// Remote Member instances must give up their own locks.
+func (ring *Ring) Unlock(key string) error {
+	info, localLock := ring.node.Locks.Load(key)
+	if !localLock {
+		return nil // lock does not exist in this local process.
+	}
+
+	lockInfo := info.(lockInfo)
+	if err := lockInfo.Lock.Unlock(lockInfo.Ctx); err != nil {
+		return errors.EnsureStack(err)
+	}
+
+	ring.node.Locks.Delete(key)
+
+	ring.Logger.WithFields(ring.lockLogFields(key)).Info("released lock")
+
+	return nil
+}
+
+// start kicks off watch(), creates a node, and renews its lease.
+// todo: replace with collections stuff.
+func (ring *Ring) start(id string) error {
```

Comment:
I think we should restructure this such that the setup is in the `New` function, but the watch functionality can be in a function and the gorutine is kicked off by the `New` function. Also, the watch should take the passed in context and respect it.

Reply:
I split the set up for testing purposes, but I can get around this by defining a different constructor in the test file.